File: app-android/app/src/main/java/com/goexercise/app/domain/RankUpDetector.kt
package com.goexercise.app.domain

/**
 * 昇格/週間ベースの達成イベント検知。iOS の `RankUpDetector`(UserDefaults 注入)の移植。
 * Android では永続層を `RankUpStore` インタフェースで抽象化し、JVM ユニットテスト可能にする。
 */
sealed class RankUpEvent {
    /** 称号ランクが上がった(to = 新ランク)。 */
    data class RankUp(val to: Int) : RankUpEvent()

    /** 連続日数が 7 の倍数に到達した(streak = その倍数)。 */
    data class Weekly(val streak: Int) : RankUpEvent()
}

/** 検知状態の永続化抽象(iOS UserDefaults 相当)。 */
interface RankUpStore {
    fun getInt(key: String): Int?
    fun putInt(key: String, value: Int)
}

class RankUpDetector(private val store: RankUpStore) {

    /**
     * 現在の連続日数からイベント列を評価し、状態を更新する。
     * - 昇格: newRank > lastRank で RankUp(newRank)。lastRank は変化があれば追従(降格も保存)。
     * - 週間: 7 の倍数 currentMultiple が 7 以上 かつ lastMultiple より大きいとき Weekly(currentMultiple)。
     */
    fun evaluate(currentStreak: Int): List<RankUpEvent> {
        val events = mutableListOf<RankUpEvent>()
        val streak = maxOf(0, currentStreak)

        // 昇格判定
        val lastRank = store.getInt(KEY_LAST_RANK) ?: 0
        val newRank = CatRank.of(streak).rank
        if (newRank > lastRank) {
            events.add(RankUpEvent.RankUp(newRank))
        }
        if (newRank != lastRank) {
            store.putInt(KEY_LAST_RANK, newRank)
        }

        // 週間倍数判定
        val currentMultiple = (streak / 7) * 7
        val lastMultiple = store.getInt(KEY_LAST_WEEKLY_MULTIPLE) ?: 0
        if (currentMultiple >= 7 && currentMultiple > lastMultiple) {
            events.add(RankUpEvent.Weekly(currentMultiple))
        }
        if (currentMultiple != lastMultiple) {
            store.putInt(KEY_LAST_WEEKLY_MULTIPLE, currentMultiple)
        }

        return events
    }


    companion object {
        private const val KEY_LAST_RANK = "rankup.lastRank"
        private const val KEY_LAST_WEEKLY_MULTIPLE = "rankup.lastWeeklyMultiple"
    }
}

File: app/GOExercise/GOExercise/Services/RankUpDetector.swift
import Foundation

/// 小節目(minor)イベント。割り込まない軽量演出のトリガ。
enum RankUpEvent: Equatable {
    case rankUp(to: Int)       // CatRank 閾値を跨いで上がった
    case weekly(streak: Int)   // 連続が新しい7の倍数に達した(大節目に未該当時)
}

/// 前回 rank と前回処理済み週次を `UserDefaults` に保持し、上昇分のみ返す純ロジック。
/// リセット(連続0)後の再上昇でも再発火する(保存値が現在値を上回ったら追従して下げる)。
struct RankUpDetector {
    private let defaults: UserDefaults
    private let rankKey = "rankup.lastRank"
    private let weeklyKey = "rankup.lastWeeklyMultiple" // 直近で発火した「7の倍数」値

    init(defaults: UserDefaults = .standard) { self.defaults = defaults }

    /// 現在の連続日数を渡し、発火すべき小節目イベント群を返す。状態は副作用で更新。
    func evaluate(currentStreak: Int) -> [RankUpEvent] {
        var events: [RankUpEvent] = []
        let streak = max(0, currentStreak)
        let newRank = CatRank(currentStreak: streak).rank
        let lastRank = defaults.object(forKey: rankKey) as? Int ?? 0

        if newRank > lastRank {
            events.append(.rankUp(to: newRank))
        }
        if newRank != lastRank {
            defaults.set(newRank, forKey: rankKey) // 上昇・下降どちらも追従
        }

        let currentMultiple = (streak / 7) * 7 // 7未満は0
        let lastMultiple = defaults.object(forKey: weeklyKey) as? Int ?? 0
        if currentMultiple >= 7, currentMultiple > lastMultiple {
            events.append(.weekly(streak: currentMultiple))
        }
        if currentMultiple != lastMultiple {
            defaults.set(currentMultiple, forKey: weeklyKey)
        }
        return events
    }

}
